resolve: Restore some effective visibility optimizations

petrochenkov opened the pull request:
Something similar was previously removed as a part of https://github.com/rust-lang/rust/pull/104602.
So we can see [bitmaps-3.1.0](https://github.com/rust-lang/rustc-perf/tree/master/collector/compile-benchmarks/bitmaps-3.1.0), [match-stress](https://github.com/rust-lang/rustc-perf/tree/master/collector/compile-benchmarks/match-stress) and [unused-warnings](https://github.com/rust-lang/rustc-perf/tree/master/collector/compile-benchmarks/unused-warnings) in regressions there, and in improvements in this PR.
After this PR all table changes should also be "locally correct" after every update.

petrochenkov commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit c2849f2e9ec7e6f43a37c59be99d723cca9dd2e1 with merge cd2317250016f7820c5a2000a8c536250d7d3dd4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c2849f2e9ec7e6f43a37c59be99d723cca9dd2e1","merge_sha":"cd2317250016f7820c5a2000a8c536250d7d3dd4"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4478429313/jobs/7873522435)
Build commit: cd2317250016f7820c5a2000a8c536250d7d3dd4 (`cd2317250016f7820c5a2000a8c536250d7d3dd4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4478429313/jobs/7873522435"},"merge_sha":"cd2317250016f7820c5a2000a8c536250d7d3dd4"} -->

rust-timer commented:
Queued cd2317250016f7820c5a2000a8c536250d7d3dd4 with parent 84c47b8279b39e165dfebeb529eb6d92592e4f8d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=84c47b8279b39e165dfebeb529eb6d92592e4f8d&end=cd2317250016f7820c5a2000a8c536250d7d3dd4).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (cd2317250016f7820c5a2000a8c536250d7d3dd4): [comparison URL](https://perf.rust-lang.org/compare.html?start=84c47b8279b39e165dfebeb529eb6d92592e4f8d&end=cd2317250016f7820c5a2000a8c536250d7d3dd4&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=84c47b8279b39e165dfebeb529eb6d92592e4f8d&end=cd2317250016f7820c5a2000a8c536250d7d3dd4&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=84c47b8279b39e165dfebeb529eb6d92592e4f8d&end=cd2317250016f7820c5a2000a8c536250d7d3dd4&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.5%  | [1.7%, 3.3%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.7% | [-4.1%, -3.3%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-2.3%, -2.3%] | 1     |
| All ❌✅ (primary)                 | -0.6% | [-4.1%, 3.3%]  | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=84c47b8279b39e165dfebeb529eb6d92592e4f8d&end=cd2317250016f7820c5a2000a8c536250d7d3dd4&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.4%, 0.4%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.7%  | [2.3%, 3.5%]   | 3     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-0.4%, 0.4%]  | 3     |
</details>

petrochenkov committed:
resolve: Restore some effective visibility optimizations

Something similar was previously removed as a part of #104602, but after this PR all table changes should also be "locally correct" after every update.

petrochenkov commented:
@bors try @rust-timer queue

bors commented:
:hourglass: Trying commit 5eda49e0d5324d7a55962b311e28bd848beef946 with merge af51eb3e6186631264bea779731f4fde52ca0be9...
<!-- homu: {"type":"TryBuildStarted","head_sha":"5eda49e0d5324d7a55962b311e28bd848beef946","merge_sha":"af51eb3e6186631264bea779731f4fde52ca0be9"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4490317922/jobs/7900001387)
Build commit: af51eb3e6186631264bea779731f4fde52ca0be9 (`af51eb3e6186631264bea779731f4fde52ca0be9`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4490317922/jobs/7900001387"},"merge_sha":"af51eb3e6186631264bea779731f4fde52ca0be9"} -->

rust-timer commented:
Queued af51eb3e6186631264bea779731f4fde52ca0be9 with parent 6502613a8199cab4cb68b08c7ddce6f210d8d69c, future [comparison URL](https://perf.rust-lang.org/compare.html?start=6502613a8199cab4cb68b08c7ddce6f210d8d69c&end=af51eb3e6186631264bea779731f4fde52ca0be9).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (af51eb3e6186631264bea779731f4fde52ca0be9): [comparison URL](https://perf.rust-lang.org/compare.html?start=6502613a8199cab4cb68b08c7ddce6f210d8d69c&end=af51eb3e6186631264bea779731f4fde52ca0be9&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=6502613a8199cab4cb68b08c7ddce6f210d8d69c&end=af51eb3e6186631264bea779731f4fde52ca0be9&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.3%] | 6     |
| Improvements ✅ <br /> (secondary) | -1.0% | [-1.7%, -0.4%] | 6     |
| All ❌✅ (primary)                 | -0.4% | [-0.4%, -0.3%] | 6     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6502613a8199cab4cb68b08c7ddce6f210d8d69c&end=af51eb3e6186631264bea779731f4fde52ca0be9&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.8% | [0.4%, 1.1%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6502613a8199cab4cb68b08c7ddce6f210d8d69c&end=af51eb3e6186631264bea779731f4fde52ca0be9&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.2% | [-4.1%, -2.4%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

petrochenkov commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit dfe2cf28bb46f0ea1a41a4c8db40ea66d8b301dd with merge 4f918dfcbe19692b255253d6c9ec331f61a279bc...
<!-- homu: {"type":"TryBuildStarted","head_sha":"dfe2cf28bb46f0ea1a41a4c8db40ea66d8b301dd","merge_sha":"4f918dfcbe19692b255253d6c9ec331f61a279bc"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4499967211/jobs/7920729466)
Build commit: 4f918dfcbe19692b255253d6c9ec331f61a279bc (`4f918dfcbe19692b255253d6c9ec331f61a279bc`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4499967211/jobs/7920729466"},"merge_sha":"4f918dfcbe19692b255253d6c9ec331f61a279bc"} -->

rust-timer commented:
Queued 4f918dfcbe19692b255253d6c9ec331f61a279bc with parent 84dd6dfd9d19176cc3c94bc1448a841e44d57890, future [comparison URL](https://perf.rust-lang.org/compare.html?start=84dd6dfd9d19176cc3c94bc1448a841e44d57890&end=4f918dfcbe19692b255253d6c9ec331f61a279bc).
<!-- rust-timer: temporary -->

bors commented:
:umbrella: The latest upstream changes (presumably #109791) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 60c6dc07dedc1c85ebfff54a0d81f3328923a2a8 with merge d14220d95ae8c760ef95a94cb4cf9b29ef0424b1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"60c6dc07dedc1c85ebfff54a0d81f3328923a2a8","merge_sha":"d14220d95ae8c760ef95a94cb4cf9b29ef0424b1"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4575402042/jobs/8080686562)
Build commit: d14220d95ae8c760ef95a94cb4cf9b29ef0424b1 (`d14220d95ae8c760ef95a94cb4cf9b29ef0424b1`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4575402042/jobs/8080686562"},"merge_sha":"d14220d95ae8c760ef95a94cb4cf9b29ef0424b1"} -->

rust-timer commented:
Queued d14220d95ae8c760ef95a94cb4cf9b29ef0424b1 with parent 22a7a19f9333bc1fcba97ce444a3515cb5fb33e6, future [comparison URL](https://perf.rust-lang.org/compare.html?start=22a7a19f9333bc1fcba97ce444a3515cb5fb33e6&end=d14220d95ae8c760ef95a94cb4cf9b29ef0424b1).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (d14220d95ae8c760ef95a94cb4cf9b29ef0424b1): [comparison URL](https://perf.rust-lang.org/compare.html?start=22a7a19f9333bc1fcba97ce444a3515cb5fb33e6&end=d14220d95ae8c760ef95a94cb4cf9b29ef0424b1&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=22a7a19f9333bc1fcba97ce444a3515cb5fb33e6&end=d14220d95ae8c760ef95a94cb4cf9b29ef0424b1&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.6%, -0.3%] | 7     |
| Improvements ✅ <br /> (secondary) | -0.6% | [-1.7%, -0.2%] | 14    |
| All ❌✅ (primary)                 | -0.4% | [-0.6%, -0.3%] | 7     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=22a7a19f9333bc1fcba97ce444a3515cb5fb33e6&end=d14220d95ae8c760ef95a94cb4cf9b29ef0424b1&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.9%  | [1.9%, 1.9%]   | 1     |
| Improvements ✅ <br /> (primary)   | -3.6% | [-3.6%, -3.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -3.6% | [-3.6%, -3.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=22a7a19f9333bc1fcba97ce444a3515cb5fb33e6&end=d14220d95ae8c760ef95a94cb4cf9b29ef0424b1&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit 60c6dc07dedc1c85ebfff54a0d81f3328923a2a8 has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco 60c6dc07dedc1c85ebfff54a0d81f3328923a2a8 -->
<!-- homu: {"type":"Approved","sha":"60c6dc07dedc1c85ebfff54a0d81f3328923a2a8","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4620157271/jobs/8170935364) [(plain)](https://github.com/rust-lang-ci/rust/commit/7a1c4f2ecbb3ab734c4d3f788733f77d86bc7f10/checks/12544148688/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:hourglass: Testing commit 60c6dc07dedc1c85ebfff54a0d81f3328923a2a8 with merge 2eaeb1eee1b21772de8b935236d16ff8e03fdcf5...
<!-- homu: {"type":"BuildStarted","head_sha":"60c6dc07dedc1c85ebfff54a0d81f3328923a2a8","merge_sha":"2eaeb1eee1b21772de8b935236d16ff8e03fdcf5"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4622737265/jobs/8177765753)
Approved by: davidtwco
Pushing 2eaeb1eee1b21772de8b935236d16ff8e03fdcf5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"davidtwco","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4622737265/jobs/8177765753"},"merge_sha":"2eaeb1eee1b21772de8b935236d16ff8e03fdcf5"} -->

rust-timer commented:
Finished benchmarking commit (2eaeb1eee1b21772de8b935236d16ff8e03fdcf5): [comparison URL](https://perf.rust-lang.org/compare.html?start=2e486be8d29d198d48bc26bfce5712a4822814f5&end=2eaeb1eee1b21772de8b935236d16ff8e03fdcf5&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=2e486be8d29d198d48bc26bfce5712a4822814f5&end=2eaeb1eee1b21772de8b935236d16ff8e03fdcf5&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.4%, -0.3%] | 6     |
| Improvements ✅ <br /> (secondary) | -0.6% | [-1.6%, -0.3%] | 12    |
| All ❌✅ (primary)                 | -0.3% | [-0.4%, -0.3%] | 6     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2e486be8d29d198d48bc26bfce5712a4822814f5&end=2eaeb1eee1b21772de8b935236d16ff8e03fdcf5&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.9%  | [1.3%, 3.1%]   | 3     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-3.1%, -1.5%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=2e486be8d29d198d48bc26bfce5712a4822814f5&end=2eaeb1eee1b21772de8b935236d16ff8e03fdcf5&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.7% | [-2.8%, -2.6%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>
